src/renderer/store/reducers/undoStackReducer.ts

```diff
@@ -0,0 +1,57 @@
+import { Reducer } from 'redux';
+import {
+  PROJECT_OPENED,
+  CURRENT_PROJECT_CLOSED,
+  UNDO_STACK_PUSHED,
+  UNDO_STACK_POPPED,
+  OP_REDONE,
+} from '../actions';
+import { Action, ApplicationStore, initialStore, Op } from '../helpers';
+import { DoPayload, UndoPayload } from '../opPayloads';
+
+const undoStackReducer: Reducer<ApplicationStore['undoStack'], Action<any>> = (
+  undoStack = initialStore.undoStack,
+  action
+) => {
+  if (
+    action.type === PROJECT_OPENED ||
+    action.type === CURRENT_PROJECT_CLOSED
+  ) {
+    return initialStore.undoStack;
+  }
+
+  if (action.type === UNDO_STACK_PUSHED) {
+    const { stack, index } = undoStack;
+
+    const newStack = stack
```

Comment:
Nice. So just to confirm, doing an action overwrites any possible redo actions from further up the stack?

Reply:
yup - otherwise we'd have an undo tree and that would get messy fast lol

Reply:
https://www.youtube.com/watch?v=D9laah5bcio&ab_channel=PatrickBrett

if you have time, this explains in detail (check the timestamps for the specific section)

Reply:
agreed.

lol didn't see vid until 1 min ago.

I drew up a diagram if you want to include in resources.

https://www.figma.com/file/ya0gV3UM0nQmXWc8SfkneQ/Design-Docs?node-id=8%3A18